[GStreamer] MediaRecorder enhancements

philn opened the pull request:
#### da92408782923545271ff4438f07eb590e3d26da
<pre>
[GStreamer] MediaRecorder enhancements
<a href="https://bugs.webkit.org/show_bug.cgi?id=281295">https://bugs.webkit.org/show_bug.cgi?id=281295</a>

Reviewed by Xabier Rodriguez-Calvar.

This patch adds WebM (for GStreamer versions &gt;= 1.24.9) and audio bitrate configuration support to
the GStreamer MediaRecorder backend. As we also handle vorbis encoding, add platform baselines for
the mimeType test which expects that codec to not be supported.

* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/glib/http/wpt/mediarecorder/MediaRecorder-first-frame-expected.txt: Added.
* LayoutTests/platform/glib/http/wpt/mediarecorder/mimeType-expected.txt: Added.
* Source/WebCore/platform/graphics/gstreamer/GStreamerRegistryScanner.cpp:
(WebCore::GStreamerRegistryScanner::isCodecSupported const):
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateGStreamer.cpp:
(WebCore::MediaRecorderPrivateBackend::MediaRecorderPrivateBackend):
(WebCore::MediaRecorderPrivateBackend::fetchData):
(WebCore::MediaRecorderPrivateBackend::containerProfile):
(WebCore::MediaRecorderPrivateBackend::configureAudioEncoder):
(WebCore::MediaRecorderPrivateBackend::preparePipeline):
(WebCore::MediaRecorderPrivateBackend::mimeType const): Deleted.
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateGStreamer.h:
(WebCore::MediaRecorderPrivateBackend::mimeType const):

Canonical link: <a href="https://commits.webkit.org/286080@main">https://commits.webkit.org/286080@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/094c0b66fbc5bcce6931d96790c742410518fb72

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/74795 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/54225 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/27612 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/79219 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/26033 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/76912 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/63361 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/2010 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/58749 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/17032 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/77862 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/48903 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/64274 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/39143 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/46173 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/21771 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/24366 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/67347 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/22116 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/80708 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/2113 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/1267 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/67010 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/2261 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/64292 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/66303 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16458 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/10248 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/8398 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/2078 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/4865 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/2106 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/3026 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/2113 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

calvaris commented on Source/WebCore/platform/mediarecorder/MediaRecorderPrivateGStreamer.cpp:
```diff
@@ -170,16 +201,16 @@ void MediaRecorderPrivateBackend::stopRecording(CompletionHandler<void()>&& comp
 
 void MediaRecorderPrivateBackend::fetchData(MediaRecorderPrivate::FetchDataCallback&& completionHandler)
 {
-    callOnMainThread([this, weakThis = ThreadSafeWeakPtr { *this }, completionHandler = WTFMove(completionHandler)]() mutable {
+    callOnMainThread([this, weakThis = ThreadSafeWeakPtr { *this }, completionHandler = WTFMove(completionHandler), mimeType = this->mimeType()]() mutable {
         auto protectedThis = weakThis.get();
         if (!protectedThis) {
-            completionHandler(nullptr, mimeType(), 0);
+            completionHandler(nullptr, mimeType, 0);
             return;
         }
         Locker locker { m_dataLock };
-        GST_DEBUG_OBJECT(m_transcoder.get(), "Transfering %zu encoded bytes", m_data.size());
+        GST_DEBUG_OBJECT(m_transcoder.get(), "Transfering %zu encoded bytes, mimeType: %s", m_data.size(), mimeType.ascii().data());
         auto buffer = m_data.take();
-        completionHandler(WTFMove(buffer), mimeType(), m_position);
+        completionHandler(WTFMove(buffer), mimeType, m_position);
     });
```
Why?

philn replied:
Why what?

calvaris replied:
I answered myself. You need to capture the mime in case the protected this is gone...

calvaris commented on Source/WebCore/platform/mediarecorder/MediaRecorderPrivateGStreamer.cpp:
```diff
@@ -328,6 +359,25 @@ void MediaRecorderPrivateBackend::setSink(GstElement* element)
+        if (options.bitsPerSecond)
+            return *options.bitsPerSecond;
+        return 0;
+    }();
```
Why convoluting this?

philn replied:
This is also how the video bitrate is figured out, I don't think it's convoluted, this pattern is used elsewhere in WebKit :)

calvaris replied:
The patter of creating a lambda with a capture that you immediately run to get a result? Might be used in places but for me it would be easier to just see a simple `if`.

philn replied:
Done!

philn committed:
[GStreamer] MediaRecorder enhancements
https://bugs.webkit.org/show_bug.cgi?id=281295

Reviewed by Xabier Rodriguez-Calvar.

This patch adds WebM (for GStreamer versions >= 1.24.9) and audio bitrate configuration support to
the GStreamer MediaRecorder backend. As we also handle vorbis encoding, add platform baselines for
the mimeType test which expects that codec to not be supported.

* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/glib/http/wpt/mediarecorder/MediaRecorder-first-frame-expected.txt: Added.
* LayoutTests/platform/glib/http/wpt/mediarecorder/mimeType-expected.txt: Added.
* Source/WebCore/platform/graphics/gstreamer/GStreamerRegistryScanner.cpp:
(WebCore::GStreamerRegistryScanner::isCodecSupported const):
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateGStreamer.cpp:
(WebCore::MediaRecorderPrivateBackend::MediaRecorderPrivateBackend):
(WebCore::MediaRecorderPrivateBackend::fetchData):
(WebCore::MediaRecorderPrivateBackend::containerProfile):
(WebCore::MediaRecorderPrivateBackend::configureAudioEncoder):
(WebCore::MediaRecorderPrivateBackend::preparePipeline):
(WebCore::MediaRecorderPrivateBackend::mimeType const): Deleted.
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateGStreamer.h:
(WebCore::MediaRecorderPrivateBackend::mimeType const):

Canonical link: https://commits.webkit.org/286080@main
